rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,355 @@
+
+1.  Each key in the `properties` of `X` is also in the `properties` of `Y`
+1.  If a property in `X` is defined as an array, then it's also defined as an array in `Y` (and therefore the same for direct references)
+1.  If a property in `X` is defined as an array with a `minItems`, then there is a `minItems` constraint on the respective property in `Y` that is equal to or larger to that in `X`
+1.  If a property in `X` is defined as an array with a `maxItems`, then there is a `maxItems` constraint on the respective property in `Y` that is less than or equal to that in `X`
+1.  Each property type referenced in `X` is compatible with the respective property type referenced in `Y`. Where respective is defined by having the same base URI (property key)
+1.  Each URI in the `required` list of `X` is also in the `required` list of `Y`
+
+#### An Array of "Property Values"
+
+Finally, the "Property Values" can be a recursive definition that infinitely nests arrays where the `items` are a `oneOf` where the options are "Property Values" variants. Because of this, compatibility is defined largely similar to above.
+
+A definition of an array of "Property Values" `X` is compatible with another definition of an array of "Property Values" `Y`, if and only if
+
+1.  Every element of `X`'s `oneOf` is compatible with at least one element of `Y`'s `oneOf`
+1.  If `X` has a `minItems` constraint specified, then there is a `minItems` constraint in `Y` that is equal to or larger to that in `X`
+1.  If `X` has a `maxItems` constraint specified, then there is a `maxItems` constraint in `Y` that is less than or equal to that in `X`
+
+### Entity Types
+
+1.  Entity types define their constraints on the structure of an entity through their `properties`, where
+    1.  The keys are base URI's to a property type
+    1.  The values are defined by either:
+        1.  A reference to a property type
+        1.  An array definition, where
+            1.  the `items` are defined by a reference to a property type
+            1.  `minItems` and/or `maxItems` are optionally defined
+1.  Entity types also define additional constraints on links from their entity through their:
+    1.  `links`, where
+        1.  The keys are _versioned URI's_ to a link type
+        1.  The values are defined by either:
+            1.  An empty object (until Link Constraints are implemented)
+            1.  An array definition, where
+                1.  the `items` are an empty object (until Link Constraints are implemented)
+                1.  `ordered` is optionally defined
+                1.  `minItems` and/or `maxItems` are optionally defined
+    1.  `requiredLinks` array, where the elements are _versioned URIs_, and is a subset of the keys in the `links`
+
+An entity type `A` is therefore compatible with another entity type `B` if and only if:
+
+1.  Each key in the `properties` of `A` is also in the `properties` of `B`
```

Comment:
Sort of the same situation here as with the property type section. Should we consider optional values on `A`'s side incompatible with `B`s that don't contain said optional values?

Reply:
Same response as above :)